### PR TITLE
server-send: use congestion_control_monitor for async writes with dyn…

### PR DESCRIPTION
**Pull request by abudnik**
…amic batch size

### CODE REVIEW

File: example/eblob_backend.cpp [hunk 1]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
 typedef std::function<int (const iterated_key_info &info)> iterator_callback;
+typedef std::function<int (std::shared_ptr<iterated_key_info> info)> iterator_server_send_callback;
 
-static iterator_callback make_iterator_server_send_callback(eblob_backend_config *c, dnet_net_state *st,
-                                                            dnet_cmd *cmd,
-                                                            ioremap::elliptics::dnet_server_send_request &request,
-                                                            uint64_t iterator_id,
-                                                            uint64_t &counter) {
+/*
+ * congestion_control_monitor is used for limiting amount of data being sent to a remote backend.
```
Comment (shaitan) [inline example/eblob_backend.cpp]:
```
 * \a congestion_control_monitor allows to limit amount of data which are sent to a remote backends simultaneously.
```

File: example/eblob_backend.cpp [hunk 2]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+ */
+class congestion_control_monitor
+{
+public:
+	/*!
+	 * Constructor: initializes internal state properly.
```
Comment (shaitan) [inline example/eblob_backend.cpp]: What does mean **properly** here? Is there any way to initialize it **improperly**?

File: example/eblob_backend.cpp [hunk 3]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+	 \a start_batch_size - maximum amount of data being sent as a first batch
+	 \a batch_timeout_msec - batch timeout in milliseconds
+	 */
+	congestion_control_monitor(uint64_t start_batch_size, int batch_timeout_msec)
+	: m_start_batch_size(start_batch_size), m_batch_size(m_start_batch_size),
```
Comment (shaitan) [inline example/eblob_backend.cpp]: Please, use one line for each initializer.

File: example/eblob_backend.cpp [hunk 4]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
-                                                            uint64_t &counter) {
+/*
+ * congestion_control_monitor is used for limiting amount of data being sent to a remote backend.
+ */
+class congestion_control_monitor
+{
+public:
+	/*!
+	 * Constructor: initializes internal state properly.
+	 \a start_batch_size - maximum amount of data being sent as a first batch
```
Comment (shaitan) [inline example/eblob_backend.cpp]: Is it limit only for the first batch, or for all batches?

File: example/eblob_backend.cpp [hunk 5]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+	 \a batch_timeout_msec - batch timeout in milliseconds
```
Comment (shaitan) [inline example/eblob_backend.cpp]: `batch timeout in milliseconds` doesn't help at all. Please, add description of its meaning. What does this parameter set in terms of 'limiting amount of data sent to a remote backends simultaneously'?

File: example/eblob_backend.cpp [hunk 6]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+	}
+
+private:
+	bool is_batch_full() const {
+		return m_bytes_pending >= m_batch_size;
+	}
+
+public:
+	/*
+	 * This constants can be used in constructor of this class.
```
Comment (shaitan) [inline example/eblob_backend.cpp]:
```
Constants which can be used in constructor of the class.
```

File: example/eblob_backend.cpp [hunk 7]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+	 */
+	static const uint64_t DEFAULT_START_BATCH_SIZE = 1024 * 1024;
```
Comment (shaitan) [inline example/eblob_backend.cpp]: Please, add description to both constants. Why does default batch size is set to 1KB?

File: example/eblob_backend.cpp [hunk 8]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+		std::unique_lock<std::mutex> lock{m_mutex};
+
+		m_bytes_processed += bytes;
+
+		if (m_bytes_processed >= m_batch_size) {
+			m_bytes_processed -= m_batch_size;
+
+			struct timeval tv;
+			gettimeofday(&tv, nullptr);
+			const int64_t elapsed = static_cast<int64_t>((tv.tv_sec - m_batch_start_time.tv_sec) * 1000 +
```
Comment (shaitan) [inline example/eblob_backend.cpp]: Why don't you use `timersub` here?

File: example/eblob_backend.cpp [hunk 9]
```diff
@@ -747,95 +748,206 @@ struct iterated_key_info {
...
+
+		auto async = session->write(info->key,
+		                            json, info->jhdr.capacity,
+		                            data, info->data_size);
 
-		async.connect([&] (const newapi::sync_write_result &/*results*/, const error_info &error) {
+		async.connect([=, &request, &monitor] (const newapi::sync_write_result &/*results*/, const error_info &error) {
 			if (st->__need_exit) {
 				dnet_backend_log(c->blog, DNET_LOG_ERROR,
 				                 "EBLOB: Interrupting server_send: peer has been disconnected");
```
Comment (shaitan) [inline example/eblob_backend.cpp]: There should be return statement after the log. Could you add it and move `monitor.remove_bytes(info->jhdr.size + info->data_size);` before `if (st->__need_exit) {`?

Comment (abudnik) [inline example/eblob_backend.cpp]: Done

File: example/eblob_backend.cpp [hunk 10]
```diff
@@ -746,96 +749,269 @@ struct iterated_key_info {
...
-				dnet_backend_log(c->blog, DNET_LOG_ERROR,
-				                 "EBLOB: server_send: %s: failed to read json: %s",
-				                 dnet_dump_id_str(info.key.id), dnet_print_error(err));
-				return 0;
-			}
-		}
+		auto serialize_response = [=, &request, &counter] (int status) {
+			return serialize(ioremap::elliptics::dnet_iterator_response{
+					iterator_id, // iterator_id
+						info->key, // key
```
Comment (shaitan) [inline example/eblob_backend.cpp]: Wrong indentions

File: example/eblob_backend.cpp [hunk 11]
```diff
@@ -746,96 +749,269 @@ struct iterated_key_info {
...
+							 "EBLOB: server_send: %s: failed to read data: %s",
+							 dnet_dump_id_str(info->key.id), dnet_print_error(err));
+					return send_fail_reply(err);
+				}
+
+				if (data_offset == 0) {
+					auto async = session->write_prepare(info->key,
+									    json, info->jhdr.capacity,
+									    data, 0, info->data_size);
+					if ((err = get_write_result(async)) != 0) {
```
Comment (shaitan) [inline example/eblob_backend.cpp]: `get_write_result` will lock current thread until result of write is received. Write by chunk should be done asynchronously without locking this thread.

File: example/eblob_backend.cpp (hunk 10 above)
Comment (abudnik) [inline example/eblob_backend.cpp]: fixed

**Comment by shaitan** [comment]
`dnet_lock_test` was failed:
```
[19:49:15][Step 1/3] # Start 9 of 14: dnet_locks_test: 
[19:49:15][Step 1/3] -- Set base directory: "/tmp/buildd/elliptics-2.26.11.0/obj-x86_64-linux-gnu/tests/result/dnet_locks_test"
[19:49:15][Step 1/3] -- Set cocaine run directory: "/tmp/elliptics-test-run-10e95fe8/"
[19:49:15][Step 1/3] Starting 1 servers
[19:49:15][Step 1/3] Starting server #1
[19:49:15][Step 1/3] Started server #1
[19:49:17][Step 1/3] Successfully connected to Cocaine #1
[19:49:17][Step 1/3] Started servers
[19:49:18][Step 1/3] Running 2 test cases...
[19:49:18][Step 1/3] /tmp/buildd/elliptics-2.26.11.0/tests/locks_test.cpp(130): fatal error in "(test_write_order_execution, use_session(n, { 1 }, 0, 0))": write_cas() failed (err=18446744073709551512): multiple consecutive writes are executed out-of-order or overlapped. Oplock mechanism of backend's request queue is broken.
[19:49:18][Step 1/3] /tmp/buildd/elliptics-2.26.11.0/tests/locks_test.cpp(161): fatal error in "(test_oplock, use_session(n, { 1 }, 0, 0))": (sess.read_data(id, 0, 0)), err: "insufficient results count due to checker: 0 of 1 (1): No such device or address: -6"
[19:49:18][Step 1/3] 
```

**Comment by abudnik** [comment]
Fixed `write_cas() failed (err=18446744073709551512)` message.
Also, -6 error is unlikely caused by changes in this PR.

File: example/eblob_backend.cpp (hunk 11 above)
Comment (shaitan) [inline example/eblob_backend.cpp]: JFYI: this was moved in separate task with performance tests.

**Pull request merged**